specification.md

```diff
@@ -454,6 +454,54 @@ to `null`.
 
 
 
+<a name="jobexecutor-list"></a>
+```java
+List<String> list()
+```
+
+Return a list of native job IDs which are known to this executor instance.  
+The returned list MAY contain IDs of jobs which were not submitted via this
+instance, and MAY be missing IDs of jobs which have been submitted by this
+instance but are finalized and purged already.  The returned job IDs MUST be
+uniquely identify job in the scope of the `JobExecutor` instance and thus in the
+scope of the backend that `JobExecutor` is bound to, but should not be assumed
+to be unique beyond that scope.
+
+IDs for any job which has been submitted via this instance and which is not yet
+in a final state MUST be returned.  Information for jobs in final state may get
+purged by the backend, and the implementation MAY purge that information also.
+The native IDs for those jobs thus MAY NOT be returned by this call, even if the
+application still holds handles to those jobs (and could thus retrieve the
+native job ID directly).
+
+
+The returned IDs can be used to re-attach a `Job` instance to the backend job
+via the `executor.attach(job, nativeJobID)` call.  This implies that the call
+SHALL only return those IDs to which the callee can attach under currently used
+authorization.
+
+
+
+<a name="jobexecutor-attach"></a>
+```java
+Job attach(Job job, String nativeJobID)
+```
+
+This method will accept a job instance in `NEW` state and a native job ID.  The
+executor will attach the Job instance to the backend job identified by that
+native (backend) job ID.  The method will return immediately, the `JobExecutor`
+will collect job state and meta data asynchronously.  A callback registered on
+the Job MUST NOT fire before the implementation completed the attachement -- at
+that point the job will have a valid `JobStatus`.  If the
+implementation is not able to attach the job (because it cannot verify
+the job ID, or the job information has been purged, etc), the job will
+be moved into `FAILED` state.
+
+The method will raise an [`InvalidJobException`](#invalidjobexception)
+if the passed job is not in `NEW` state.  Any job status, resource spec
+etc. which are set on the passed `Job` instance will be discarded.
```

Comment:
Sorry for nitpicking, but should we mandate that this be so or warn that this could be so?

Reply:
IMHO we should mandate, and I changed the text: MUST raise, MUST discard.  If we don't discard any conflicting attributes, we end up with an inconsistent set of information which may change over time, which is not good.  We could also mandate that the application did not set any attributes on the Job instance before calling - that would work for me also and might be cleaner?

Reply:
@hategan : ping

Reply:
Sorry. Yeah, I see your point.

I'm considering two separate scenarios:
1. in which users may want to provide a persistence layer that fully restores job objects to a live state. To do so, if the executor does not guarantee that the job spec/info will be restored, the user must.
2. users can still set the wrong info after attach() leading to the same problem; so I don't think the issue is fully enforceable unless we somehow make certain properties to be write-once.

I'm writing this because I've kinda ran into such situations. But please consider this conversation resolved from my perspective (aka. press the "Resolve conversation" button unless you want to discuss further).

Reply:
>     2\. users can still set the wrong info after attach()

I don't mind giving the user that freedom (and responsibility),  So, let's leave as it is (`MUST be NEW, attach MUST discard`), and reconsider once it turns out to be problematic (which I doubt).  Either way, thanks for the nitpick.